src/engagement_db_to_coda/engagement_db_to_coda.py

```diff
@@ -0,0 +1,184 @@
+    :type coda_dataset_config: src.engagement_db_to_coda.configuration.CodaDatasetConfiguration
+    :param db_message: Message to add to Coda.
+    :type db_message: engagement_database.data_models.Message
+    """
+    log.debug("Adding message to Coda")
+
+    coda_message = core_data_modules.data_models.Message(
+        message_id=db_message.coda_id,
+        text=db_message.text,
+        creation_date_time_utc=TimeUtils.datetime_to_utc_iso_string(db_message.timestamp),
+        labels=[]
+    )
+
+    # If the engagement database message already has labels, initialise with these in Coda.
+    if len(db_message.labels) > 0:
+        # TODO: Validate that these labels are valid under the code schemes being copied to.
+        coda_message.labels = db_message.labels
+
+    # Otherwise, run any auto-coders that are specified.
+    else:
+        for scheme_config in coda_dataset_config.code_scheme_configurations:
+            if scheme_config.auto_coder is None:
+                continue
+            label = CleaningUtils.apply_cleaner_to_text(scheme_config.auto_coder, db_message.text, scheme_config.code_scheme)
+            if label is not None:
+                coda_message.labels.append(label)
+
+    # Add the message to the Coda dataset.
+    coda.add_message(coda_dataset_config.coda_dataset_id, coda_message)
+
+
+def _update_db_message_from_coda_message(engagement_db, db_message, coda_message, coda_config, transaction=None):
+    """
+    Updates a message in the engagement database based on the labels in the Coda message.
+
+    If the labels match, returns without updating anything.
+    Otherwise, if the new labels contain a WS code, clears the labels and updates the dataset.
+    Otherwise, overwrites the existing labels with the new labels.
+
+    :param engagement_db: Engagement database to update the message in.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param db_message: Engagement database message to update
+    :type db_message: engagement_database.data_models.Message
+    :param coda_message: Coda message to use to update the engagement database message.
+    :type coda_message: core_data_modules.data_models.Message
+    :param coda_config: Configuration for the update.
+    :type coda_config:  src.engagement_db_to_coda.configuration.CodaSyncConfiguration
+    :param transaction: Transaction in the engagement database to perform the update in.
+    :type transaction: google.cloud.firestore.Transaction | None
+    """
+    coda_dataset_config = coda_config.get_dataset_config_by_engagement_db_dataset(db_message.dataset)
+
+    # Check if the labels in the engagement database message already match those from the coda message.
+    # If they do, return without updating anything.
+    # TODO: Validate if the dataset is correct too?
+    if json.dumps([x.to_dict() for x in db_message.labels]) == json.dumps([y.to_dict() for y in coda_message.labels]):
```

Comment:
Can we add dunder eq to core_data_modules.data_models.Label?

Reply:
Added here https://github.com/AfricasVoices/CoreDataModules/pull/220, will update here once reviewed